sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -181,3 +187,239 @@ case class Randn(child: Expression, hideSeed: Boolean = false) extends RDG {
+    Seq(RUNTIME_REPLACEABLE, EXPRESSION_WITH_RANDOM_SEED)
+
+  override val dataType: DataType = {
+    val first = min.dataType
+    val second = max.dataType
+    (min.dataType, max.dataType) match {
+      case _ if !valid(min) || !valid(max) => NullType
+      case (_, LongType) | (LongType, _) if Seq(first, second).forall(integer) => LongType
+      case (_, IntegerType) | (IntegerType, _) if Seq(first, second).forall(integer) => IntegerType
+      case (_, ShortType) | (ShortType, _) if Seq(first, second).forall(integer) => ShortType
+      case (_, DoubleType) | (DoubleType, _) => DoubleType
+      case (_, FloatType) | (FloatType, _) => FloatType
+      case _ => NullType
+    }
+  }
+
+  private def valid(e: Expression): Boolean = e.dataType match {
+    case _: ShortType | _: IntegerType | _: LongType | _: FloatType | _: DoubleType => true
+    case _ => false
+  }
+
+  private def integer(t: DataType): Boolean = t match {
+    case _: ShortType | _: IntegerType | _: LongType => true
+    case _ => false
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    var result: TypeCheckResult = TypeCheckResult.TypeCheckSuccess
+    def requiredType = "integer or floating-point"
+    Seq((min, "min", 0),
+      (max, "max", 1),
+      (seed, "seed", 2)).foreach {
+      case (expr: Expression, name: String, index: Int) =>
+        if (!expr.foldable && result == TypeCheckResult.TypeCheckSuccess) {
+          result = DataTypeMismatch(
+            errorSubClass = "NON_FOLDABLE_INPUT",
+            messageParameters = Map(
+              "inputName" -> name,
+              "inputType" -> requiredType,
+              "inputExpr" -> toSQLExpr(expr)))
+        } else if (!valid(expr) && result == TypeCheckResult.TypeCheckSuccess) {
+          result = DataTypeMismatch(
+            errorSubClass = "UNEXPECTED_INPUT_TYPE",
+            messageParameters = Map(
+              "paramIndex" -> ordinalNumber(index),
+              "requiredType" -> requiredType,
+              "inputSql" -> toSQLExpr(expr),
+              "inputType" -> toSQLType(expr.dataType)))
+        }
+    }
+    result
+  }
+
+  override def left: Expression = min
+  override def right: Expression = max
+
+  override def seedExpression: Expression = seed
+  override def withNewSeed(newSeed: Long): Expression = {
+    val result = Uniform(min, max)
+    result.seed = Literal(newSeed, LongType)
+    result
+  }
+
+  override def withNewChildrenInternal(newFirst: Expression, newSecond: Expression): Expression =
+    Uniform(newFirst, newSecond)
+
+  override def replacement: Expression = {
+    def cast(e: Expression, to: DataType): Expression = if (e.dataType == to) e else Cast(e, to)
+    cast(Add(
+      cast(min, DoubleType),
+      Multiply(
+        Subtract(
+          cast(max, DoubleType),
+          cast(min, DoubleType)),
+        Rand(seed))),
+      dataType)
+  }
+}
+
+object Uniform {
+  lazy val random = new Random()
```

Comment:
Why don't you follow the existing approach for the random expression, and don't extend `RDG`?

Reply:
Good question, `RDG` `extends UnaryExpression`, whereas this new function accepts two arguments.

Reply:
Update on this: I was able to split the `UnaryExpression` part of `RDG` into a separate class and now `Uniform` can reuse the remainder of `RDG` to simplify this.